File: lib/test/structuredfield/parseInnerList.test.ts
import assert from 'node:assert';
import test from 'node:test';
import { SfItem } from '../../src/structuredfield/SfItem.js';
import { parseInnerList } from '../../src/structuredfield/parse/parseInnerList.js';
import { parseList } from '../../src/structuredfield/parse/parseList.js';

test('parseInnerList', () => {
	assert.deepStrictEqual(parseInnerList(`( 1 2 3 )`), {
		value: new SfItem([1, 2, 3]),
		src: ``,
	});
	assert.deepStrictEqual(parseInnerList(`(1)`), {
		value: new SfItem([1]),
		src: ``,
	});
	assert.deepStrictEqual(parseInnerList(`()`), {
		value: new SfItem([]),
		src: ``,
	});
	assert.deepStrictEqual(parseList(`(1 1.23 a "a" ?1 :AQID: @1659578233)`), {
		value: [
			new SfItem([
				1,
				1.23,
				Symbol.for('a'),
				'a',
				true,
				new Uint8Array([1, 2, 3]),
				new Date(1659578233 * 1000),
			]),
		], src: ``,
	}
	);
	assert.throws(() => parseInnerList(`[1 2 3)`), /failed to parse "\[1 2 3\)" as Inner List/);
	assert.throws(() => parseInnerList(`(1 2 3]`), /failed to parse "\]" as Inner List/);
	assert.throws(() => parseInnerList(`(`), /failed to parse "" as Inner List/);
});

File: python/src/test/structuredfield/parseInnerList_test.py
import unittest
from datetime import datetime

from ...structuredfield.parse.ParsedValue import ParsedValue
from ...structuredfield.parse.parseInnerList import parseInnerList
from ...structuredfield.SfException import SfException
from ...structuredfield.SfItem import SfItem
from ...structuredfield.SfToken import SfToken


class TestDecodeSfItem(unittest.TestCase):
    def test_parseInnerList(self):
        self.assertEqual(
            parseInnerList("( 1 2 3 )"),
            ParsedValue(SfItem([SfItem(1), SfItem(2), SfItem(3)]), ""),
        )
        self.assertEqual(
            parseInnerList("(1)"),
            ParsedValue(SfItem([SfItem(1)]), ""),
        )
        self.assertEqual(
            parseInnerList("()"),
            ParsedValue(SfItem([]), ""),
        )
        self.assertEqual(
            parseInnerList('(1 1.23 a "a" ?1 :AQID: @1659578233)'),
            ParsedValue(
                SfItem(
                    [
                        SfItem(1),
                        SfItem(1.23),
                        SfItem(SfToken("a")),
                        SfItem("a"),
                        SfItem(True),
                        SfItem(b"\x01\x02\x03"),
                        SfItem(datetime.fromtimestamp(1659578233)),
                    ]
                ),
                "",
            ),
        )

        self.assertRaises(SfException, parseInnerList, "[1 2 3)")
        self.assertRaises(SfException, parseInnerList, "(1 2 3]")
        self.assertRaises(SfException, parseInnerList, "(")
